docs/src/development.md

```diff
@@ -1,33 +1,103 @@
+and trust each other.
+
+Core contributors have one more right that others do not -- they can create repository branches.
+Thus, they can contribute "locally" - by pushing proposed changed to the primary repository's work branches,
+and create "local" pull requests inside the same repo.  This is different from others,
+who can contribute only from their own forks.
+
+There is little difference between creating pull requests from one's own fork and from the primary repo,
+and there are a few reasons why core contributors should **never** do it from the primary repo:
+
+* it ensures that casual contributors always run the same CI as core contributors. If contribution process breaks, it will affect everyone, and will get fixed faster.
+* it puts everyone on the same leveled playing field, reducing the "caste system" effect, making the project feel more welcoming to new contributors
+* it ensures that the primary repo only has maintained branches (e.g. `main` and `v1.x`),
+  not a bunch of PR branches whose ownership and work status is unclear to everyone
+
+In the martin repository, we follow this and have a branch protection rule that prevents core contributors from creating pull requests from the primary repo.
+
+</details>
+
+```bash
+# clone the primary fork to your local machine, naming the remote "upstream"
+# make sure to replace the URL with the correct one
+git clone -o upstream https://github.com/maplibre/martin.git
 cd martin
+
+# add your own fork as a remote, naming it "origin"
+git remote add origin https://github.com/nyurik/martin.git
 ```
 
-Fork Martin repo into your own GitHub account, and add your fork as a remote
+For further setup instructions for IDEs, please see the [Getting Involved](getting-involved.md) step after you have installed the necessary tools below.
+<details><summary>If you have already cloned the repo locally, use this guide to update your setup (click to expand)</summary>
 
-```bash, ignore
-git remote add origin  _URL_OF_YOUR_FORK_
+If you already cloned the repo locally, you can update it to use the new setup. This assumes you have a local clone of the repo, the remote name is `origin`, and you have already forked the repo on GitHub.
+
+```bash
+# Rename the existing remote to "upstream". Your "main" branch will now track "upstream/main"
+git remote rename origin upstream
+
+# Add your own fork as a remote, naming it "origin" (adjust the URL)
+git remote add origin https://github.com/nyurik/martin.git
 ```
 
-Install [docker](https://docs.docker.com/get-docker/) and [docker-compose](https://docs.docker.com/compose/)
+</details>
 
-```bash, ignore
-# Ubuntu-based distros have an older version that might also work:
-sudo apt install -y  docker.io docker-compose
+### Contributing New Code
+
+```bash
+# switch to main branch (tracking upstream/main), and pull the latest changes
+git switch main
+git pull
+
+# create a new branch for your work
+git switch -c my-new-feature
```

Comment:
If a new contributor needs to read this, then git pull is dangerous for she/him. git fetch is always better.

```bash
git fetch upstream
git switch -c my-new-feature upstream/main
```